Userland/Libraries/LibGfx/ImageFormats/AVIFLoader.cpp

```diff
@@ -0,0 +1,202 @@
+
+static ErrorOr<void> decode_avif_header(AVIFLoadingContext& context)
+{
+    if (context.state >= AVIFLoadingContext::HeaderDecoded)
+        return {};
+
+    avifDecoder* decoder = avifDecoderCreate();
+    if (decoder == NULL) {
+        return Error::from_string_literal("failed to allocate AVIF decoder");
+    }
+
+    avifResult result = avifDecoderSetIOMemory(decoder, context.data.data(), context.data.size());
+    if (result != AVIF_RESULT_OK) {
+        avifDecoderDestroy(decoder);
+        return Error::from_string_literal("Cannot set IO on avifDecoder");
+    }
+
+    result = avifDecoderParse(decoder);
+    if (result != AVIF_RESULT_OK) {
+        avifDecoderDestroy(decoder);
+        return Error::from_string_literal("Failed to decode AVIF");
+    }
+
+    // image header decoded, save some results in context
+    context.size = IntSize { decoder->image->width, decoder->image->height };
+    context.has_alpha = decoder->alphaPresent == 1;
+    context.image_count = decoder->imageCount;
+    context.repetition_count = decoder->repetitionCount <= 0 ? 0 : decoder->repetitionCount;
+    context.state = AVIFLoadingContext::State::HeaderDecoded;
+
+    avifDecoderDestroy(decoder);
+    return {};
+}
+
+static ErrorOr<void> decode_avif_image(AVIFLoadingContext& context)
+{
+    VERIFY(context.state >= AVIFLoadingContext::State::HeaderDecoded);
+
+    avifDecoder* decoder = avifDecoderCreate();
+    if (decoder == NULL)
+        return Error::from_string_literal("failed to allocate AVIF decoder");
+
+    avifResult result = avifDecoderSetIOMemory(decoder, context.data.data(), context.data.size());
+    if (result != AVIF_RESULT_OK) {
+        avifDecoderDestroy(decoder);
+        return Error::from_string_literal("Cannot set IO on avifDecoder");
+    }
+
+    result = avifDecoderParse(decoder);
+    if (result != AVIF_RESULT_OK) {
+        avifDecoderDestroy(decoder);
+        return Error::from_string_literal("Failed to decode AVIF");
+    }
+
+    avifRGBImage rgb;
+    while (avifDecoderNextImage(decoder) == AVIF_RESULT_OK) {
+        auto bitmap_format = context.has_alpha ? BitmapFormat::BGRA8888 : BitmapFormat::BGRx8888;
+        auto bitmap = TRY(Bitmap::create(bitmap_format, context.size.value()));
+
+        avifRGBImageSetDefaults(&rgb, decoder->image);
+        rgb.format = avifRGBFormat::AVIF_RGB_FORMAT_BGRA;
+
+        result = avifRGBImageAllocatePixels(&rgb);
+        if (result != AVIF_RESULT_OK) {
+            avifDecoderDestroy(decoder);
+            return Error::from_string_literal("Allocation of RGB samples failed");
+        }
+
+        result = avifImageYUVToRGB(decoder->image, &rgb);
+        if (result != AVIF_RESULT_OK) {
+            avifDecoderDestroy(decoder);
+            avifRGBImageFreePixels(&rgb);
+            return Error::from_string_literal("Conversion from YUV failed");
+        }
+
+        // FIXME: support high bitdepth
+        VERIFY(rgb.depth == 8);
```

Comment:
Does this cause crashes when decoding 10- or 12-bit images? Perhaps it would be better to return an error here rather than crashing.

Reply:
Yes, this does. I've changed it to an error. It would be better to try to render them, though. Does anyone know of `Gfx::Bitmap` can support this?

Reply:
Not in its current form

Reply:
We would have to convert it down to 8-bit with tonemapping, but currently that code lives in LibMedia, so it would take some fiddling to make it work here. IMO, it'd be better to wait until we can move image codecs somewhere other than LibGfx (whether it be a small library just for them that depends on LibMedia, or whatever else).

Reply:
I've added a check to error out in the high bit depth cases for now, along with a matching test.